File: Other/Review/Phase 5/Review1/disjoint-set/disjoint-set.go
package main

import "fmt"

type UnionFind struct {
	root []int
}

func New(size int) *UnionFind {
	root := make([]int, size)

	for i := 0; i < size; i++ {
		root[i] = i
	}

	return &UnionFind{
		root: root,
	}
}

func (uf *UnionFind) Find(x int) int {
	if uf.root[x] == x {
		return x
	}
	return uf.Find(uf.root[x])
}

func (uf *UnionFind) Union(x int, y int) {
	root_x := uf.Find(x)
	root_y := uf.Find(y)

	if root_x != root_y {
		uf.root[root_y] = root_x
	}
}

func (uf *UnionFind) IsConnected(x int, y int) bool {
	return uf.Find(x) == uf.Find(y)
}

func main() {
	unionFind := New(10)
	unionFind.Union(1, 2)
	unionFind.Union(2, 3)
	unionFind.Union(4, 5)
	unionFind.Union(6, 7)
	unionFind.Union(8, 9)
	unionFind.Union(1, 9)
	fmt.Println(unionFind.IsConnected(1, 9))
	fmt.Println(unionFind.IsConnected(3, 9))
	fmt.Println(unionFind.IsConnected(1, 4))
	fmt.Println(unionFind.IsConnected(8, 4))
	fmt.Println(unionFind.IsConnected(4, 5))
	fmt.Println(unionFind.IsConnected(5, 6))
	fmt.Println(unionFind.IsConnected(6, 7))
}

File: Other/Review/Stage 10/Review 2/Part1/disjoint-set/disjoint-set.ts
export class DisjointSet {
  private root: number[] = [];

  constructor(size: number) {
    this.root = new Array(size).fill(0).map((_, index) => {
      return index;
    });
  }

  find(x: number): number {
    if (this.root[x] === x) {
      return x;
    }
    return this.find(this.root[x]);
  }

  union(x: number, y: number) {
    const root_x = this.root[x];
    const root_y = this.root[y];

    if (root_x !== root_y) {
      this.root[root_y] = root_x;
    }
  }

  isConnected(x: number, y: number): boolean {
    return this.find(x) === this.find(y);
  }
}

// test
const unionFind = new DisjointSet(10);
unionFind.union(1, 2);
unionFind.union(2, 3);
unionFind.union(4, 5);
unionFind.union(6, 7);
unionFind.union(8, 9);
unionFind.union(1, 9);
console.log(unionFind.isConnected(1, 9));
console.log(unionFind.isConnected(3, 9));
console.log(unionFind.isConnected(1, 4));
console.log(unionFind.isConnected(8, 4));
console.log(unionFind.isConnected(4, 5));
console.log(unionFind.isConnected(5, 6));
console.log(unionFind.isConnected(6, 7));
